README.md
# :nut_and_bolt: vue-plugin-boilerplate

[![CircleCI](https://circleci.com/gh/kazupon/vue-plugin-boilerplate/tree/master.svg?style=svg)](https://circleci.com/gh/kazupon/vue-plugin-boilerplate/tree/master)

Boilerplate for Vue.js plugin


## :rocket: Usage

    $ npm i -g vue-cli
    $ vue init kazupon/vue-plugin-boilerplate vue-hello-plugin

It’s up to you !!

## :hammer: Using tools
- Compiler
    - TypeScript
- Transpile
    - babel (for development)
    - buble (for distribution)
- Linter
    - eslint
- Bundler
    - webpack (for development)
    - rollup (for distribution)
- Test Assertion
    - power-assert
- Test Framework
    - mocha
- Test Runner
    - karma
- Test Coverage
    - istanbul
- Headless Browser
    - chrome headless
- End-to-End Test Fremework
    - nightwatch.js

## :package: Bunlded ModuleName
Defined the came-case into global.

ex:
```
- `name` of package.json
vue-plugin-boilerplate

- bundled
VuePluginBoilerplate

- javascript usage example
Vue.use(VuePluginBoilerplate)
```

# :copyright: License

[MIT](http://opensource.org/licenses/MIT)


CHANGELOG.md
<a name="20.0.0"></a>
# [20.0.0](https://github.com/kazupon/vue-plugin-boilerplate/compare/v19.0.0...v20.0.0) (2018-09-24)


### :boom: Breaking changes

* back to webpack-dev-server ([60696de](https://github.com/kazupon/vue-plugin-boilerplate/commit/60696de))



<a name="19.0.0"></a>
# [19.0.0](https://github.com/kazupon/vue-plugin-boilerplate/compare/v18.0.0...v19.0.0) (2018-05-28)


### :boom: Breaking changes

* **docs:** switch to vuepress from gitbook ([435432f](https://github.com/kazupon/vue-plugin-boilerplate/commit/435432f))



<a name="18.0.0"></a>
# [18.0.0](https://github.com/kazupon/vue-plugin-boilerplate/compare/v17.1.0...v18.0.0) (2018-05-28)


### :boom: Breaking changes

* remove flow setup ([7157fb5](https://github.com/kazupon/vue-plugin-boilerplate/commit/7157fb5))
* switch typescript ([cb8f934](https://github.com/kazupon/vue-plugin-boilerplate/commit/cb8f934))


### :zap: Improvements

* upgrade babel 7 ready ([c946d30](https://github.com/kazupon/vue-plugin-boilerplate/commit/c946d30))
* upgrade vue releated templates ([98b3123](https://github.com/kazupon/vue-plugin-boilerplate/commit/98b3123))



<a name="17.1.0"></a>
# [17.1.0](https://github.com/kazupon/vue-plugin-boilerplate/compare/v17.0.0...v17.1.0) (2017-09-08)


### :star: New Features

* add cross-env ([5fb74fb](https://github.com/kazupon/vue-plugin-boilerplate/commit/5fb74fb))



<a name="17.0.0"></a>
# [17.0.0](https://github.com/kazupon/vue-plugin-boilerplate/compare/v16.0.0...v17.0.0) (2017-07-01)


### :boom: Breaking changes

* swtich to chrome headless from phantomjs ([d2ef3ad](https://github.com/kazupon/vue-plugin-boilerplate/commit/d2ef3ad))



<a name="16.0.0"></a>
# [16.0.0](https://github.com/kazupon/vue-plugin-boilerplate/compare/v15.0.1...v16.0.0) (2017-06-07)


### :boom: Breaking changes

* **lint:** switch eslint-config-vue to eslint-coneslint-plugin-vue-libs ([433ec30](https://github.com/kazupon/vue-plugin-boilerplate/commit/433ec30))



<a name="15.0.1"></a>
## [15.0.1](https://github.com/kazupon/vue-plugin-boilerplate/compare/v15.0.0...v15.0.1) (2017-04-01)


### :zap: Improvements

* **test:** add initial cb ([82e738c](https://github.com/kazupon/vue-plugin-boilerplate/commit/82e738c))



<a name="15.0.0"></a>
# [15.0.0](https://github.com/kazupon/vue-plugin-boilerplate/compare/v14.0.0...v15.0.0) (2017-03-27)


### :boom: Breaking changes

* **test:** switch jasmine to mocha ([f2f7539](https://github.com/kazupon/vue-plugin-boilerplate/commit/f2f7539))


### :zap: Improvements

* **test:** ignore vue performance info ([6d94f34](https://github.com/kazupon/vue-plugin-boilerplate/commit/6d94f34))



<a name="14.0.0"></a>
# [14.0.0](https://github.com/kazupon/vue-plugin-boilerplate/compare/v13.1.1...v14.0.0) (2017-03-23)


### :star: New Features

* **build:** support es module building ([577938d](https://github.com/kazupon/vue-plugin-boilerplate/commit/577938d)), closes [#10](https://github.com/kazupon/vue-plugin-boilerplate/issues/10)
* **config:** add unpkg hint ([0593890](https://github.com/kazupon/vue-plugin-boilerplate/commit/0593890)), closes [#11](https://github.com/kazupon/vue-plugin-boilerplate/issues/11)
* **gitbook:** add getting started template ([2d14c84](https://github.com/kazupon/vue-plugin-boilerplate/commit/2d14c84))



<a name="13.1.1"></a>
## [13.1.1](https://github.com/kazupon/vue-plugin-boilerplate/compare/v13.1.0...v13.1.1) (2017-02-16)


### :bug: Bug Fixes

* **circleci:** fix typo ([8675a8d](https://github.com/kazupon/vue-plugin-boilerplate/commit/8675a8d))


### :zap: Improvements

* **gitbook:** ignore gitbook docs additional versioning ([5ac0d6d](https://github.com/kazupon/vue-plugin-boilerplate/commit/5ac0d6d))



<a name="13.1.0"></a>
# [13.1.0](https://github.com/kazupon/vue-plugin-boilerplate/compare/v13.0.1...v13.1.0) (2017-02-16)


### :star: New Features

* add babel-polyfill ([a56db01](https://github.com/kazupon/vue-plugin-boilerplate/commit/a56db01))
* doc module versioning ([bbcb6bc](https://github.com/kazupon/vue-plugin-boilerplate/commit/bbcb6bc))


### :zap: Improvements

* **eslint:** update eslint ([a6230e9](https://github.com/kazupon/vue-plugin-boilerplate/commit/a6230e9))
* **npm:** update tasks ([c0fac2d](https://github.com/kazupon/vue-plugin-boilerplate/commit/c0fac2d))
* **test:** more easy unit testing ([d29e70c](https://github.com/kazupon/vue-plugin-boilerplate/commit/d29e70c))



<a name="13.0.1"></a>
## [13.0.1](https://github.com/kazupon/vue-plugin-boilerplate/compare/v13.0.0...v13.0.1) (2017-02-04)


### :zap: Improvements

* **meta:** github account default username ([da4886d](https://github.com/kazupon/vue-plugin-boilerplate/commit/da4886d))



<a name="13.0.0"></a>
# [13.0.0](https://github.com/kazupon/vue-plugin-boilerplate/compare/v12.0.0...v13.0.0) (2017-02-03)


### :zap: Improvements

* **power-assert:** change babel-preset-power-assert ([d8e5f9c](https://github.com/kazupon/vue-plugin-boilerplate/commit/d8e5f9c))
* **webpack:** upgrade webpack 2.2 ([f1dcdb8](https://github.com/kazupon/vue-plugin-boilerplate/commit/f1dcdb8))


<a name="12.0.0"></a>
# [12.0.0](https://github.com/kazupon/vue-plugin-boilerplate/compare/v11.2.2...v12.0.0) (2016-12-28)


### :star: New Features

* add badges ([fe8b4d1](https://github.com/kazupon/vue-plugin-boilerplate/commit/fe8b4d1))
* add ci service setup ([f692bd3](https://github.com/kazupon/vue-plugin-boilerplate/commit/f692bd3))
* add editor config ([2273cd9](https://github.com/kazupon/vue-plugin-boilerplate/commit/2273cd9))
* support covrage service selecting ([c52059d](https://github.com/kazupon/vue-plugin-boilerplate/commit/c52059d))


### :zap: Improvements

* **coverage:** change to istanbul from coverage for babel plugin ([f7e4cf9](https://github.com/kazupon/vue-plugin-boilerplate/commit/f7e4cf9))
* **saucelabs:** update mobile envs ([8d3f882](https://github.com/kazupon/vue-plugin-boilerplate/commit/8d3f882))



<a name="11.2.2"></a>

config/karma.conf.js
// Karma configuration
// Generated on Tue Sep 08 2015 19:27:24 GMT+0900 (JST)

module.exports = function (config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',

    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['mocha'],

    // list of files / patterns to load in the browser
    files: [
      '../test/**/*.js'
    ],

    // list of files to exclude
    exclude: [
    ],

    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      '../src/**/*.js': ['webpack', 'sourcemap'],
      '../test/**/*.js': ['webpack', 'sourcemap']
    },

    webpack: {
      devtool: 'inline-source-map',
      module: {
        loaders: [{
          test: /\.js$/,
          exclude: /node_modules/,
          loader: 'babel',
          query: {
            presets: ['es2015'],
            plugins: ['babel-plugin-espower']
          }
        }],
        postLoaders: [{
          test: /\.json$/,
          loader: 'json'
        }, {
          test: /\.js$/,
          exclude: /test|node_modules/,
          loader: 'istanbul-instrumenter'
        }]
      }
    },

    webpackMiddleware: {
      noInfo: true
    },

    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: [
      'mocha', 'coverage'
    ],

    coverageReporter: {
      reporters: [{
        type: 'lcov', dir: '../coverage'
      }, {
        type: 'text-summary', dir: '../coverage'
      }]
    },

    // web server port
    port: 9876,

    // enable / disable colors in the output (reporters and logs)
    colors: true,

    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,

    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,

    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['PhantomJS'],

    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: true
  })
}
